web/crux/src/domain/agent.spec.ts

```diff
@@ -0,0 +1,380 @@
+import { Subject, firstValueFrom, skip, Observable, finalize } from 'rxjs'
+import { CloseReason } from 'src/grpc/protobuf/proto/agent'
+import { CruxPreconditionFailedException } from 'src/exception/crux-exception'
+import {
+  ContainerCommandRequest,
+  ContainerOperation,
+  DeleteContainersRequest,
+  DeploymentStatus,
+  DeploymentStatusMessage,
+  Empty,
+} from 'src/grpc/protobuf/proto/common'
+import { DeploymentStatusEnum } from '@prisma/client'
+import { Agent, AgentEvent } from './agent'
+
+const AGENT_ID = 'agent-id'
+const AGENT_ADDRESS = '127.0.0.1:1234'
+const AGENT_VERSION = 'agent-version'
+const AGENT_PUBLIC_KEY = 'public-key'
+
+const GrpcNodeConnectionMock = jest.fn().mockImplementation(() => ({
+  nodeId: AGENT_ID,
+  address: AGENT_ADDRESS,
+  connectedAt: new Date(),
+}))
+
+const observableComplete = <T>(it: Observable<T>) =>
+  new Promise(resolve => {
+    it.pipe(finalize(() => resolve(null))).subscribe()
+  })
+
+it('containerPrefixNameOf should return the combined prefix name', () => {
+  expect(
+    Agent.containerPrefixNameOf({
+      prefix: null,
+      name: 'name',
+    }),
+  ).toEqual('name')
+
+  expect(
+    Agent.containerPrefixNameOf({
+      prefix: 'prefix',
+      name: 'name',
+    }),
+  ).toEqual('prefix-name')
+})
+
+describe('agent', () => {
+  let agentConnection: any = null
+  let agent: Agent = null
+  let eventChannel: Subject<AgentEvent> = null
+
+  beforeEach(() => {
+    eventChannel = new Subject<AgentEvent>()
+
+    agentConnection = new GrpcNodeConnectionMock()
+
+    agent = new Agent(
+      agentConnection,
+      {
+        id: AGENT_ID,
+        version: AGENT_VERSION,
+        publicKey: AGENT_PUBLIC_KEY,
+      },
+      eventChannel,
+    )
+  })
+
+  describe('connection status', () => {
+    it('onConnected should send connection status', async () => {
+      const eventPromise = firstValueFrom(eventChannel)
+
+      agent.onConnected()
+
+      await expect(eventPromise).resolves.toEqual({
+        id: AGENT_ID,
+        address: AGENT_ADDRESS,
+        status: 'connected',
+        version: AGENT_VERSION,
+        connectedAt: agentConnection.connectedAt,
+        updating: false,
+      })
+    })
+
+    it('onDisconnected should send unreachable status', async () => {
+      const eventPromise = firstValueFrom(eventChannel.pipe(skip(1)))
```

Comment:
```
const resultObs = result.pipe(toArray())
const actual = await firstValueFrom(resultObs)
```

Reply:
`toArray` requires the subject to complete which in this case it never does. Manually completing it works but it requires more code to do the same thing.

```
it('onConnected should send connection status', async () => {
      const eventPromise = firstValueFrom(eventChannel.pipe(toArray()))

      agent.onConnected()

      eventChannel.complete()
      await expect(eventPromise).resolves.toEqual({
        id: AGENT_ID,
        address: AGENT_ADDRESS,
        status: 'connected',
        version: AGENT_VERSION,
        connectedAt: agentConnection.connectedAt,
        updating: false,
      })
    })
```